### ISSUE TITLE
None changes annotation behaviour

### ISSUE THREAD

**Issue by munkybutt** [microsoft/pyright#4604]
**Describe the bug**
When annotating a variable returned from a function with no annotations, the variable annotation is ignored.
This only seems to be the behaviour if the return type is `Unknown | None`, `Unknown` allows the annotation.
I cannot add annotations to the function as it is from a third party package.

**To Reproduce**
This first example has the following setup:
`third_party_module` is obscured, as it is a module exposing c++ functions with no annotations available.
`return_one` is in a third party module written in python, so I cannot add annotations.
`return_two_from_one` is my own function, so I can annotate the variable returned from `return_one` as `int`.
This works as expected, `one` is defined as `int` once assigned.
```python
import third_party_module  # type: ignore

def return_one():
    one = third_party_module.one()
    return one

def return_two_from_one():
    one: int = return_one()
    two = one + 1
    return two

two = return_two_from_one()
```
![image](https://user-images.githubusercontent.com/7100563/217778916-0dabee27-4aa7-4dc8-b735-cdeddda928e1.png)

The second example has the introduction of `None` as a potential return type, so `Unknown | None`.
This then ignores the type annotation for `one`, which is `int | None` and maintains `Unknown | None`.
This seems inconsistent that `None` causes the annotation to be ignored.
```python
import third_party_module  # type: ignore

def return_one():
    one = third_party_module.one()
    if one is None:
        return
    return one

def return_two() -> int:
    one: int | None = return_one()
    assert one is not None
    two = one + 1
    return two

two = return_two()
```
![image](https://user-images.githubusercontent.com/7100563/217780711-47a24764-20f7-4b6c-84a9-921481aefd84.png)

**Expected behavior**
The type of `one` in the second example to evaluate to `int | None`.

This may be expected, but I think this behaviour has changed as I have been working with this codebase for along time and have been annotating variables in this manner since the beginning. I cannot be certain though as I am unable to rollback to earlier versions to compare.

**VS Code extension or command-line**
Using Pylance 2023.1.40 (25 January 2023) Release

**Comment by erictraut** [comment]
The type checker is doing the correct thing here and is properly narrowing the type of `one` from `None | Unknown` to `Unknown` by virtue of the `assert one is not None` expression. However, the hover provider is not displaying the type `Unknown` as it should. This appears to be due to a regression that was introduced recently. @bschnurr, it looks like you changed the code in the hover provider in this [part of the code](https://github.com/microsoft/pyright/blob/e7e19af5793e2b1e4cb5f2c9cdd177f64eb4ac43/packages/pyright-internal/src/languageService/hoverProvider.ts#L231). I don't understand what the change was meant to do (there is no associated comment), but it is now ignoring the type returned by `getType` if the evaluated type is `Unknown`. Looks like a bug. Could you please investigate?

**Comment by erictraut** [comment]
@rchiodo, Bill looked into this and said that he thinks this regression was introduced as part of your recent pytest changes. Could you take a look at it when you have a chance?

**Comment by rchiodo** [comment]
Will do.

**Comment by rchiodo** [comment]
Just to be clear, the bug here is this:

![image](https://user-images.githubusercontent.com/19672699/218599509-e93ab719-8d99-4978-a1a1-c42d1f4f8e7e.png)

This was introduced with the pytest support.

It should look like this:

![image](https://user-images.githubusercontent.com/19672699/218599751-3d3b4d17-2495-4547-8e2d-4ca258828c92.png)

Which is what the 2022.1.30 (and earlier) version shows.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.294, which I just published. It will also be included in a future release of pylance.
